src/repo.rs

```diff
@@ -1585,16 +1639,12 @@ impl Repo {
                 info.start_pending_removal();
             }
 
-            // Poll close sinks, and remove those that are ready or errored.
-            let sinks_to_close: Vec<RepoId> = self
-                .remote_repos
-                .drain()
-                .map(|(repo_id, remote_repo)| {
-                    let pending = self.pending_close_sinks.entry(repo_id.clone()).or_default();
-                    pending.push(remote_repo.sink);
-                    repo_id
-                })
-                .collect();
+            // close all open sinks
+            for (repo_id, remote_repo) in self.remote_repos.drain() {
+                let pending_sinks = self.pending_close_sinks.entry(repo_id.clone()).or_default();
+                pending_sinks.push(remote_repo.sink);
+            }
+            let sinks_to_close = self.pending_close_sinks.keys().cloned().collect::<Vec<_>>();
```

Comment:
I'm not sure if this is a good idea, because perhaps not all sinks pending close will be ready to be polled? In the current code, only the currently "open" sinks are marked as pending close, and then `poll_close` is called on them. The ones already inside `pending_close_sinks` are only polled again in response to `WakeSignal::PendingCloseSink(sink_id)` in the below loop.

Reply:
Good spot. I actually had this in here because it made the tests pass and in the process of trying to explain why it made the tests  pass I realised I had some logic elsewhere slightly wrong. The reason I had to add this was that the tests sometimes hung waiting for sinks to close. What was happening was this scenario:


* The repo has a connection to a remote
* The test calls `new_remote_repo` with the same `RepoId` as the existing remote
* The test then calls `RepoHandle::shutdown`
* This results in the events `ConnectRemoteRepo` and `Stop` being received consecutively
* The repo handles `ConnectRemoteRepo`, notices that it is a connect event for an existing stream and adds the `Sink` corresponding to the existing connection to `pending_close_sinks` _but fails to call `Self::poll_close_sinks`_
* Then the `Stop` event is received and we transition into the shutdown routine.
* The shutdown routine adds the new sink to `pending_close_sinks`, then calls `poll_close_sinks` on the new sink
* Now we wait for all the sinks to close, _but nothing every called poll_close on the old sink we replaced when handing `ConnectRemoteRepo`


I've updated the code to do closing sinks in a helper method which always calls `Self::poll_close_sinks`

Reply:
(In case it wasn't clear, I've reverted the logic here to the original "only poll the sinks which are ready" logic because I fixed the bug described above)

Reply:
Ok I see, thanks for the fix.